llvm/lib/SYCLLowerIR/SYCLPropagateAspectsUsage.cpp

```diff
@@ -441,6 +448,96 @@ void createUsedAspectsMetadataForFunctions(
   }
 }
 
+void createConditionallyUsedAspectsMetadataForFunctions(
+    FuncToConditionExprsAndConditionalAspectsTy
+        &FunctionToConditionExpressionsAndConditionalAspects,
+    const AspectsSetTy &ExcludeAspectVals) {
+  for (auto &[F, ConditionsAndAspectsVec] :
+       FunctionToConditionExpressionsAndConditionalAspects) {
+    LLVMContext &C = F->getContext();
+
+    // Create a set of unique conditions and aspects. First we add the ones from
+    // the found aspects that have not been excluded.
+    ConditionExprsAndConditionalAspectsTy UniqueConditionsAndAspects;
+    for (const auto &ConditionsAndAspectsPair : ConditionsAndAspectsVec) {
+      AspectsSetTy FilteredAspects;
+      for (const auto &A : ConditionsAndAspectsPair.second)
+        if (!ExcludeAspectVals.contains(A))
+          FilteredAspects.insert(A);
+
+      if (FilteredAspects.empty())
+        continue;
+
+      UniqueConditionsAndAspects.push_back(
+          std::make_pair(ConditionsAndAspectsPair.first, FilteredAspects));
+    }
+
+    // If there is new metadata, merge it with the old aspects. We preserve
+    // the excluded ones.
+
+    if (const MDNode *ExistingAspects =
+            F->getMetadata("sycl_conditionally_used_aspects")) {
+      for (const MDOperand &MDOp : ExistingAspects->operands()) {
+        if (MDNode *PairNode = dyn_cast<MDNode>(MDOp)) {
+          if (PairNode->getNumOperands() != 2)
+            continue;
+
+          MDNode *ConditionNode = dyn_cast<MDNode>(PairNode->getOperand(0));
+          ConditionsSetTy ConditionsSet;
+          if (ConditionNode) {
```

Comment:
I think that `dyn_cast` on 485 should just be `cast`. Therefore, we should be able to drop the `if` here.

It is us who controls the format and I think that we can rely on it being correct.

Similar comment applies to the code below about aspects metadata operand.